driver.go

```diff
@@ -64,94 +62,230 @@ func (d *Driver) Close() {
+// 	if err != nil {
+// 		return nil, 0, err
+// 	}
+// 	if ret.Next() {
+// 		record := ret.Record()
+// 		if ret.Next() {
+// 			return nil, 0, ErrMultipleFound(ID)
+// 		}
+// 		node, ok := record.Values[0].(neo4j.Node)
+// 		if !ok {
+// 			return nil, 0, ls.ErrNotFound(ID)
+// 		}
+// 		return s.MakeNode(node), node.Id, nil
+// 	}
+// 	return nil, 0, ls.ErrNotFound(ID)
+// }
 
-// LoadNode attempts to load a node with the given ID. The ID is
-// looked up in id node property
-func (s *Session) LoadNode(tx neo4j.Transaction, labels []string, ID string) (ls.Node, int64, error) {
-	query := fmt.Sprintf("MATCH (node%s {%s:$id}) RETURN node", makeLabels(labels), PropNodeID)
-	data := map[string]interface{}{"id": ID}
-	s.Logf("Query: %s with %v", query, data)
-	ret, err := tx.Run(query, data)
-	if err != nil {
-		return nil, 0, err
+// // MakeNode builds a graph node from the given db node
+// func (s *Session) MakeNode(node neo4j.Node) ls.Node {
+// 	retNode := ls.NewNode(fmt.Sprint(node.Props[PropNodeID]), node.Labels...)
+// 	if x, ok := node.Props[PropNodeValue]; ok {
+// 		if str, ok := x.(string); ok {
+// 			retNode.SetValue(str)
+// 		}
+// 	}
+// 	buildPropertyMap(node.Props, retNode.GetProperties())
+// 	return retNode
+// }
+
+// func buildPropertyMap(in map[string]interface{}, out map[string]*ls.PropertyValue) {
+// 	for k, v := range in {
+// 		switch k {
+// 		case PropNodeValue:
+// 		case PropNodeID:
+// 		default:
+// 			switch val := v.(type) {
+// 			case []interface{}:
+// 				arr := make([]string, 0, len(val))
+// 				for _, x := range val {
+// 					arr = append(arr, fmt.Sprint(x))
+// 				}
+// 				out[k] = ls.StringSlicePropertyValue(arr)
+// 			default:
+// 				out[k] = ls.StringPropertyValue(fmt.Sprint(val))
+// 			}
+// 		}
+// 	}
+// }
+
+// CreateGraph creates a graph and returns the neo4j ID of the root node
+func CreateGraph(session *Session, tx neo4j.Transaction, nodes []graph.Node) (int64, error) {
+	nodeIds := make(map[graph.Node]int64)
+	allNodes := make(map[graph.Node]struct{})
+	entityNodes := make(map[graph.Node]struct{})
+	// Find triples:
+	for _, node := range nodes {
+		allNodes[node] = struct{}{}
 	}
-	if ret.Next() {
-		record := ret.Record()
-		if ret.Next() {
-			return nil, 0, ErrMultipleFound(ID)
+OUTER:
+	for node := range allNodes {
+		hasEdges := false
+
+		//var en []graph.Node
+		//var entityRootId int64
+		if _, exists := node.GetProperty(ls.EntitySchemaTerm); exists {
+			var root graph.Node
+			fmt.Println(session.emptyDB(tx))
+			if _, err := session.existsDB(tx, node); err != nil && !session.emptyDB(tx) {
+				fmt.Println("HERE")
+				// entityNodes[node] = struct{}{}
+				// entityRootId = dbRoot.Id
+				// fmt.Println(dbRoot.Id)
+				// fmt.Println(dbRoot.Props)
+				// nodeIds[node] = id
+
+				root = node
+
+				ls.IterateDescendants(root, func(nd graph.Node, _ []graph.Node) bool {
+					if _, exists := nd.GetProperty(ls.EntityIDTerm); exists {
```

Comment:
All you need to do here is to collect all nodes under the root node. Keep in mind, if you are here, then:
  * The node is an entity root node
  * The node exists in the db
 So, you should build a set of nodes that should be excluded from the `nodes` array. This can simply be done by:

```
excludeNodes:=make(map[graph.Node]struct{})
...
ls.IterateDescendants(root,func(nd graph.Node,_[]graph.Node) bool {
  if nd!=root { // Do not exclude the root node.
      excludeNodes[nd]=struct{}{}
  }
 return true
}, ...
```
Once you have this, you should process triples if:
  * Both from and to nodes of the edge are in the `nodes` array (which you already do)
  * Both from and to nodes of the edge are not in `excludeNodes`
 This will create all nodes that are not in db, and add all necessary edges (including those involving the entity root node).